### PR TITLE
Deploy features/yoshi_pins

### PR DESCRIPTION
**Pull request by zhangchuck**
- Implemented parts of #71 for intial testing.

# Finished

- `!put`/`!p` command
- `!get`/`!g` command
- basic persistence on the docker container

# Still to-do

- [x] Backing up the database to S3 (skeleton added)
- [x] Retrieving the database on restart of docker (skeleton added)
- [x] List all existing pins (need to discuss UI)
- [x] Figure out why the `quote-bot-dev` is not picking up the `!g` `!p` and `!l` commands
~~- [ ] Improve the matching for the `!list` command (skeleton added)~~
- [x] Create an S3 bucket
~~- [ ] Generate credentials (`aws_access_key_id` and `aws_secret_access_key`)~~
~~- [ ] Pass credentials to docker environment~~

Notes on credential passing:

``` the IAM role is in the EC2 instance metadata, some googling suggest that docker containers on the instance have access to that metadata already```

### CODE REVIEW

File: discord_quote/discord_quote/discord_quote.py [hunk 1]
```diff
@@ -517,6 +575,199 @@ def pred(m):
...
+    aliases = db_execute(f"SELECT alias FROM pins WHERE alias = \"{alias}\";")
+    if len(aliases) > 0:
+        log.info(log_msg(['sent_message',
+                          'invalid_pin_request',
+                          ctx.message.channel.name,
+                          'Alias too long.']))
+        await ctx.send(f'*{alias}* has already been used as a pin alias')
+        return
+
+    ### ALSO NEED TO ADD A CONDITIONAL FOR AN ALREADY USED ALIAS
```
Comment (zhangchuck) [inline discord_quote/discord_quote/discord_quote.py:652]: Line to be deleted

**Comment by zhangchuck** [comment]
Note that this feature is not working on `quote-bot-dev` yet. We'll need to inspect the logs at some point.

**Comment by zhangchuck** [comment]
Issue with deployment was that we ran out of diskspace and the image wouldn't download properly. To fix this, we make sure to prune docker every time we redeploy:

play4honor/p4h_webhooks@b8d2675

**Comment by zhangchuck** [comment]
Remember to pull the hardcoded bucket name and place it into an environment variable

**Comment by zhangchuck** [comment]
- [ ] Improve the matching for the `!list` command (skeleton added) will be moved to its own issue.

File: discord_quote/discord_quote/discord_quote.py [hunk 2]
```diff
@@ -517,6 +609,237 @@ def pred(m):
...
+    if '\r' in msg_target or '\n' in msg_target:
+      # If weird users decide to separate the msg_id from the reply using a line return
+      # clean it up.
+      if '\r' in msg_target:
+        _temp = msg_target.split('\r')
+      else:
+        _temp = msg_target.split('\n')
+
+      msg_target = _temp[0].strip()
+      alias = ' '.join([_temp[1].strip()] + request.split(' ')[1:]).lower()
```
Comment (synsypa) [inline discord_quote/discord_quote/discord_quote.py:650]: does this produce a key error for `_temp[1]` if you dont provide an alias

File: discord_quote/discord_quote/discord_quote.py [hunk 3]
```diff
@@ -517,6 +609,237 @@ def pred(m):
...
+
+    # Check if alias already exists
+    ### MAYBE WE SHOULD JUST SET A PRIMARY KEY IN THE SCHEMA AND HANDLE THE
+    ### SQLITE ERROR
+    aliases = db_execute(f"SELECT alias FROM pins WHERE alias = \"{alias}\";")
+    if len(aliases) > 0:
+        log.info(log_msg(['sent_message',
+                          'invalid_pin_request',
+                          ctx.message.channel.name,
+                          'Alias too long.']))
```
Comment (synsypa) [inline discord_quote/discord_quote/discord_quote.py:682]: this error message is inaccurate

File: discord_quote/discord_quote/discord_quote.py [hunk 4]
```diff
@@ -517,6 +609,237 @@ def pred(m):
...
+                      msg_id,
+                      ctx.guild.get_channel(channel_id).name,
+                      msg_.author.name,
+                      msg_.created_at.strftime("%Y-%m-%d %H:%M:%S"),
+                      ctx.message.author.name,
+                      msg_.clean_content]))
+
+        # Store the message
+        row = [alias, msg_.jump_url, ctx.message.author.name, ctx.message.created_at]
+        db_execute(
```
Comment (synsypa) [inline discord_quote/discord_quote/discord_quote.py:723]: we probalby want a log that this insert was successful

File: discord_quote/discord_quote/discord_quote.py [hunk 5]
```diff
@@ -517,6 +609,237 @@ def pred(m):
...
+        log.info(log_msg(['sent_message',
+                          'invalid_pin_request',
+                          ctx.message.channel.name]))
+
+@bot.command(aliases=['g'])
+async def get(ctx, *, request:str):
+    pin = db_execute(
+            f"SELECT msg_url FROM pins WHERE alias=\"{request.lower()}\""
+    )
+    if len(pin) > 0:
```
Comment (synsypa) [inline discord_quote/discord_quote/discord_quote.py:787]: this probably could be `==1` as extra protection against doubles?

File: discord_quote/discord_quote/discord_quote.py [hunk 6]
```diff
@@ -517,6 +609,237 @@ def pred(m):
...
+            f"Couldn't find ({msg_id}) from channel {channel_id}. " +
+            f"Requested by {ctx.message.author.name}."
+        )
+
+        log.info(log_msg(['sent_message',
+                          'invalid_pin_request',
+                          ctx.message.channel.name]))
+
+@bot.command(aliases=['g'])
+async def get(ctx, *, request:str):
```
Comment (synsypa) [inline discord_quote/discord_quote/discord_quote.py:783]: `request` should probably be `alias` just so the !help docstring isnt useless

File: discord_quote/discord_quote/discord_quote.py (hunk 5 above)
Comment (zhangchuck) [reply discord_quote/discord_quote/discord_quote.py:790]: I want this to still work even when there are duplicates. Rare case where I want a silent failure.

**Review by synsypa — APPROVED** [review]

**Pull request merged**